File: day15/src/main.py
import sys
from pathlib import Path

DIRECTIONS = {'^': (-1, 0), 'v': (1, 0), '<': (0, -1), '>': (0, 1)}

TRANSLATE = {'#': '##', 'O': '[]', '.': '..', '@': '@.'}


def load_data(file_path, is_wide=False):
    lines = Path(file_path).read_text().strip().splitlines()
    layout = []
    moves = ''
    robot_pos = None
    boxes = set()

    for line in lines:
        if line.startswith('#'):
            layout.append(list(''.join(TRANSLATE[char] if is_wide else char for char in line)))
        elif line.startswith(tuple(DIRECTIONS.keys())):
            moves += line.strip()

    for r, row in enumerate(layout):
        for c, char in enumerate(row):
            if char == '@':
                robot_pos = (r, c)
                layout[r][c] = '.'
            elif char == 'O':
                boxes.add(frozenset([(r, c)]))
                layout[r][c] = '.'
            elif char == '[':
                boxes.add(frozenset([(r, c), (r, c + 1)]))
                layout[r][c] = layout[r][c + 1] = '.'

    return layout, robot_pos, boxes, moves


from collections import deque


def move(layout, boxes, r, c, dr, dc):
    rows, cols = len(layout), len(layout[0])
    target_r, target_c = r + dr, c + dc

    if target_r < 0 or target_r >= rows or target_c < 0 or target_c >= cols or layout[target_r][target_c] == '#':
        return r, c

    queue, moved = deque([(target_r, target_c)]), set()

    while queue:
        curr_r, curr_c = queue.popleft()
        for box in {b for b in boxes if (curr_r, curr_c) in b and b not in moved}:
            new_positions = {(br + dr, bc + dc) for br, bc in box}
            if any(nr < 0 or nr >= rows or nc < 0 or nc >= cols or layout[nr][nc] == '#' for nr, nc in new_positions):
                return r, c
            queue.extend(new_positions - {(curr_r, curr_c)})
            moved.add(box)

    if moved:
        boxes.difference_update(moved)
        boxes.update(frozenset((br + dr, bc + dc) for br, bc in box) for box in moved)

    return target_r, target_c


def execute(layout, robot_pos, boxes, moves):
    for i, move_dir in enumerate(moves):
        if i % (len(moves) // 100 or 1) == 0:
            sys.stdout.write(f'\rProgress {i * 100 // len(moves)}%')
            sys.stdout.flush()

        robot_pos = move(layout, boxes, *robot_pos, *DIRECTIONS[move_dir])

    sys.stdout.write('\r\033[2K')
    return robot_pos, boxes


def calculate_gps(boxes):
    return sum(100 * min(r for r, _ in box) + min(c for _, c in box) for box in boxes)


def part1():
    layout, robot_pos, boxes, moves = load_data('my_input.txt', is_wide=False)
    robot_pos, boxes = execute(layout, robot_pos, boxes, moves)
    print(f'Answer: {calculate_gps(boxes)}')


def part2():
    layout, robot_pos, boxes, moves = load_data('my_input.txt', is_wide=True)
    robot_pos, boxes = execute(layout, robot_pos, boxes, moves)
    print(f'Answer: {calculate_gps(boxes)}')


if __name__ == '__main__':
    part1()
    part2()

File: day15/src/main.rs
use std::collections::{HashMap, HashSet, VecDeque};
use std::fs;
use std::io::{self, Write};
use std::sync::LazyLock;

const DIRECTIONS: &[(char, (isize, isize))] =
    &[('^', (-1, 0)), ('v', (1, 0)), ('<', (0, -1)), ('>', (0, 1))];

static TRANSLATE: LazyLock<HashMap<char, &'static str>> = LazyLock::new(|| {
    HashMap::from([
        ('#', "##"),
        ('O', "[]"),
        ('.', ".."),
        ('@', "@."),
    ])
});

fn parse_input(
    filename: &str,
    is_wide: bool,
) -> (
    Vec<Vec<char>>,
    (usize, usize),
    HashSet<Vec<(usize, usize)>>,
    String,
) {
    let data = fs::read_to_string(filename).expect("Failed to read file");
    let mut layout = Vec::new();
    let mut moves = String::new();
    let mut robot_pos = None;
    let mut boxes = HashSet::new();

    for line in data.trim().lines() {
        if line.starts_with('#') {
            let translated: Vec<char> = if is_wide {
                line.chars()
                    .flat_map(|c| {
                        TRANSLATE
                            .get(&c)
                            .map(|&v| v.to_string()) // Map to `String`
                            .unwrap_or_else(|| c.to_string()) // Use fallback as `String`
                            .chars()
                            .collect::<Vec<_>>()
                    })
                    .collect()
            } else {
                line.chars().collect()
            };
            layout.push(translated);
        } else if DIRECTIONS.iter().any(|(d, _)| line.starts_with(*d)) {
            moves.push_str(line.trim());
        }
    }

    for (r, row) in layout.iter().enumerate() {
        for (c, &char) in row.iter().enumerate() {
            match char {
                '@' => { robot_pos = Some((r, c)); }
                'O' => { boxes.insert(vec![(r, c)]); }
                '[' => { boxes.insert(vec![(r, c), (r, c + 1)]); }
                _ => {}
            }
        }
    }

    (layout, robot_pos.unwrap(), boxes, moves)
}

fn move_robot(
    layout: &[Vec<char>],
    boxes: &mut HashSet<Vec<(usize, usize)>>,
    r: usize,
    c: usize,
    dr: isize,
    dc: isize,
) -> (usize, usize) {
    let rows = layout.len();
    let cols = layout[0].len();
    let target_r = (r as isize + dr) as usize;
    let target_c = (c as isize + dc) as usize;

    if target_r >= rows || target_c >= cols || layout[target_r][target_c] == '#' {
        return (r, c);
    }

    let mut queue = VecDeque::from([(target_r, target_c)]);
    let mut moved = HashSet::new();

    while let Some((curr_r, curr_c)) = queue.pop_front() {
        let current_boxes: Vec<_> = boxes
            .iter()
            .filter(|box_vec| box_vec.contains(&(curr_r, curr_c)) && !moved.contains(*box_vec))
            .cloned()
            .collect();

        for box_vec in current_boxes {
            let mut new_positions = Vec::new();
            for &(br, bc) in &box_vec {
                let nr = (br as isize + dr) as usize;
                let nc = (bc as isize + dc) as usize;
                if nr >= rows || nc >= cols || layout[nr][nc] == '#' {
                    return (r, c);
                }
                new_positions.push((nr, nc));
            }
            queue.extend(&new_positions);
            moved.insert(box_vec.clone());
        }
    }

    boxes.retain(|b| !moved.contains(b));
    boxes.extend(moved.into_iter().map(|b| {
        b.into_iter()
            .map(|(br, bc)| ((br as isize + dr) as usize, (bc as isize + dc) as usize))
            .collect()
    }));

    (target_r, target_c)
}

fn execute(
    layout: &[Vec<char>],
    mut robot_pos: (usize, usize),
    boxes: &mut HashSet<Vec<(usize, usize)>>,
    moves: &str,
) -> (usize, usize) {
    let len = moves.len();
    for (i, move_dir) in moves.chars().enumerate() {
        if i % (len / 100).max(1) == 0 {
            print!("\rProgress: {}%", i * 100 / len);
            io::stdout().flush().unwrap();
        }

        let (dr, dc) = DIRECTIONS.iter().find(|&&(d, _)| d == move_dir).unwrap().1;
        robot_pos = move_robot(layout, boxes, robot_pos.0, robot_pos.1, dr, dc);
    }

    print!("\r\x1b[2K");

    robot_pos
}

fn calculate_gps(boxes: &HashSet<Vec<(usize, usize)>>) -> usize {
    boxes
        .iter()
        .map(|box_vec| {
            let min_r = box_vec.iter().map(|&(r, _)| r).min().unwrap();
            let min_c = box_vec.iter().map(|&(_, c)| c).min().unwrap();
            100 * min_r + min_c
        })
        .sum()
}

fn part1() {
    let (layout, robot_pos, mut boxes, moves) = parse_input("my_input.txt", false);
    let _final_robot_pos = execute(&layout, robot_pos, &mut boxes, &moves);
    let gps_sum = calculate_gps(&boxes);

    println!("Answer: {}", gps_sum);
}

fn part2() {
    let (layout, robot_pos, mut boxes, moves) = parse_input("my_input.txt", true);
    let _final_robot_pos = execute(&layout, robot_pos, &mut boxes, &moves);
    let gps_sum = calculate_gps(&boxes);

    println!("Answer: {}", gps_sum);
}

fn main() {
    part1();
    part2();
}
